pkg/pillar/cmd/volumemgr/handlevolume.go

```diff
@@ -0,0 +1,148 @@
+	configArg interface{}) {
+
+	log.Infof("handleVolumeCreate(%s)", key)
+	config := configArg.(types.VolumeConfig)
+	ctx := ctxArg.(*volumemgrContext)
+	updateVolume(ctx, config)
+	log.Infof("handleVolumeCreate(%s) Done", key)
+}
+
+func handleVolumeModify(ctxArg interface{}, key string,
+	configArg interface{}) {
+
+	log.Infof("handleVolumeModify(%s)", key)
+	config := configArg.(types.VolumeConfig)
+	ctx := ctxArg.(*volumemgrContext)
+	updateVolume(ctx, config)
+	log.Infof("handleVolumeModify(%s) Done", key)
+}
+
+func handleVolumeDelete(ctxArg interface{}, key string,
+	configArg interface{}) {
+
+	log.Infof("handleVolumeDelete(%s)", key)
+	config := configArg.(types.VolumeConfig)
+	ctx := ctxArg.(*volumemgrContext)
+	deleteVolume(ctx, config)
+	log.Infof("handleVolumeDelete(%s) Done", key)
+}
+
+func publishVolumeStatus(ctx *volumemgrContext,
+	status *types.VolumeStatus) {
+
+	key := status.Key()
+	log.Debugf("publishVolumeStatus(%s)", key)
+	pub := ctx.pubVolumeStatus
+	pub.Publish(key, *status)
+	log.Debugf("publishVolumeStatus(%s) Done", key)
+}
+
+func unpublishVolumeStatus(ctx *volumemgrContext,
+	status *types.VolumeStatus) {
+
+	key := status.Key()
+	log.Debugf("unpublishVolumeStatus(%s)", key)
+	pub := ctx.pubVolumeStatus
+	c, _ := pub.Get(key)
+	if c == nil {
+		log.Errorf("unpublishVolumeStatus(%s) not found", key)
+		return
+	}
+	pub.Unpublish(key)
+	log.Debugf("unpublishVolumeStatus(%s) Done", key)
+}
+
+func lookupVolumeStatus(ctx *volumemgrContext,
+	key string) *types.VolumeStatus {
+
+	log.Infof("lookupVolumeStatus(%s)", key)
+	pub := ctx.pubVolumeStatus
+	c, _ := pub.Get(key)
+	if c == nil {
+		log.Infof("lookupVolumeStatus(%s) not found", key)
+		return nil
+	}
+	status := c.(types.VolumeStatus)
+	log.Infof("lookupVolumeStatus(%s) Done", key)
+	return &status
+}
+
+func updateVolume(ctx *volumemgrContext,
+	config types.VolumeConfig) {
+
+	log.Infof("updateVolume for %v", config.VolumeID)
+	status := lookupVolumeStatus(ctx, config.Key())
+	if status == nil {
+		status = &types.VolumeStatus{
+			VolumeID:                config.VolumeID,
+			ContentID:               config.ContentID,
+			VolumeContentOriginType: config.VolumeContentOriginType,
+			MaxVolSize:              config.MaxVolSize,
+			GenerationCounter:       config.GenerationCounter,
+			DisplayName:             config.DisplayName,
+			ReadOnly:                config.ReadOnly,
+			RefCount:                config.RefCount,
+			LastUse:                 time.Now(),
+		}
+	}
+	publishVolumeStatus(ctx, status)
+	changed, _ := doUpdateVol(ctx, status)
+	if changed {
+		publishVolumeStatus(ctx, status)
+	}
+	log.Infof("updateVolume for %v Done", config.VolumeID)
+}
+
+func deleteVolume(ctx *volumemgrContext,
+	config types.VolumeConfig) {
+
+	log.Infof("deleteVolume for %v", config.VolumeID)
+	status := lookupVolumeStatus(ctx, config.Key())
+	if status == nil {
+		log.Infof("deleteVolume for %v, VolumeStatus not found", config.VolumeID)
+		return
+	}
```

Comment:
Don't you need to do a refcount decrement and check if it reached zero here?

Reply:
No, unpublishing of volume config from the zedagent will be published only if it is removed from the device configuration. And, if the volume is removed from the device config then we don't need to check for refcount.

Reply:
OK, but that makes the checks in the GC code different because refcount==0 means it should still stay around, unless the VolumeConfig is deleted. Yet you don't have a VolumeConfig at boot and you have a refcount=0 VolumeStatus. Hence I don't understand what the invarant is for when the volume should exist.
Please explain.

FWIW to me it is easier to have the existence in the device API mean that the refcount is at least one. Then if N app instances use the volume the refcount=1+N. Then if it is deleted from the device API the refcount=N. Then as all of the app instances using the volume have been halted the refcount drops to zero and it is deleted.
Thus the existence of a volume with refcount=0 is only a transient condition during boot until the know device configuration has been processed by all the agents.

Reply:
Added. Decrementing and checking the refcount before deleting the volume.